external/storm-autocreds/src/main/java/org/apache/storm/common/AbstractAutoCreds.java

```diff
@@ -0,0 +1,250 @@
+                if (!res.containsKey(configKey) && configMap.containsKey(configKey)) {
+                    res.put(configKey, configMap.get(configKey));
+                }
+            }
+        }
+        return res;
+    }
+
+    @Override
+    public void renew(Map<String, String> credentials, Map topologyConf) {
+        doRenew(credentials, updateConfigs(topologyConf));
+    }
+
+    @Override
+    public void populateCredentials(Map<String, String> credentials) {
+        credentials.put(getCredentialKey(StringUtils.EMPTY),
+                DatatypeConverter.printBase64Binary("dummy place holder".getBytes()));
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void populateSubject(Subject subject, Map<String, String> credentials) {
+        addCredentialToSubject(subject, credentials);
+        addTokensToUGI(subject);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void updateSubject(Subject subject, Map<String, String> credentials) {
+        addCredentialToSubject(subject, credentials);
+        addTokensToUGI(subject);
+    }
+
+    protected Set<Pair<String, Credentials>> getCredentials(Map<String, String> credentials) {
+        Set<Pair<String, Credentials>> res = new HashSet<>();
+        if (configKeys != null) {
+            for (String configKey : configKeys) {
+                Credentials cred = doGetCredentials(credentials, configKey);
+                if (cred != null) {
+                    res.add(new Pair(configKey, cred));
+                }
+            }
+        } else {
+            Credentials cred = doGetCredentials(credentials, StringUtils.EMPTY);
+            if (cred != null) {
+                res.add(new Pair(StringUtils.EMPTY, cred));
+            }
+        }
+        return res;
+    }
+
+    protected void fillHadoopConfiguration(Map topoConf, String configKey, Configuration configuration) {
+        Map<String, Object> config = (Map<String, Object>) topoConf.get(configKey);
+        LOG.info("TopoConf {}, got config {}, for configKey {}", topoConf, config, configKey);
+        if (config != null) {
+            List<String> resourcesToLoad = new ArrayList<>();
+            for (Map.Entry<String, Object> entry : config.entrySet()) {
+                if (entry.getKey().equals(CONFIG_KEY_RESOURCES)) {
+                    resourcesToLoad.addAll((List<String>) entry.getValue());
+                } else {
+                    configuration.set(entry.getKey(), String.valueOf(entry.getValue()));
+                }
+            }
+            LOG.info("Resources to load {}", resourcesToLoad);
+            // add configs from resources like hdfs-site.xml
+            for (String pathStr : resourcesToLoad) {
+                configuration.addResource(new Path(Paths.get(pathStr).toUri()));
+            }
+        }
+        LOG.info("Initializing UGI with config {}", configuration);
+        UserGroupInformation.setConfiguration(configuration);
```

Comment:
Never mind. It seems that this will always be executed in sequence for hdfs and hbase. So it will be overwriting the configuration and then login and get the tokens.